### ISSUE TITLE
add hyper-v scsi (gen2) support

### ISSUE THREAD

**Issue by eyberg** [nanovms/nanos#1416]
https://github.com/nanovms/ops/issues/547

nanos currently works on hyper-v gen1 and it looks like we are using that on azure as well:

```
azure_image.go:                 HyperVGeneration: compute.HyperVGenerationTypesV1,
```

scsi should be much faster

**Comment by am11** [comment]
Some notes from https://community.spiceworks.com/topic/451467-hyperv-scsi-controller-v-ide-controller?page=1#entry-3877743

#### Gen 1

The IDE controller is emulated (vmwp.exe) where as the SCSI controller is synthetic.

The IDE controller implements an emulated IDE controller which means there is extra processing before the I/O is sent to the disk.

The SCSI controller in Windows Server virtualization is not emulated, instead it uses the VMBUS (Virtual Machine BUS). These devices have an advantage over emulated devices because they leverage the existing Windows driver architecture.

With SCSI, we can hot-swap a drive, resize (incl. shrink), without shutting down the VM, which is not possible with IDE.

It seems, we cannot boot with gen 1 SCSI drives.

#### Gen 2

Supports UEFI and therefore supports native VMBUS on boot. "Gen2" VM's, SCSI and IDE both, utilize VMBUS therefore there is no performance gains on "Gen2" VM's in using either.

Not sure about hot-swapping or resizing of live drive with Gen 2 IDE, but SCSI has this ability with both generations.

---

Verdict: `add hyper-v scsi (gen2) support` (aka title of this issue) is the way to enhance HyperV support! 😎

**Comment by francescolavra** [comment]
SCSI is actually supported already by Nanos, which can use SCSI volumes just fine. The problem is that hyper-v gen1 VMs cannot be configured to boot from SCSI, because the hyper-v BIOS supports only IDE as bootable hard disks. So SCSI disks in hyper-v gen1 VMS can only be used as additional (non-bootable) volumes, while the boot disk must be IDE.

In order to support gen2 VMS (which use UEFI and can boot from SCSI disks), we need to add support for UEFI-based booting to the Nanos bootloader.

**Comment by am11** [comment]
Ah, makes sense. Should we re-purpose this issue for UEFI support (regardless of target provider)?

Seems like the [RH's shim](https://github.com/rhboot/shim) (signed by MSFT's key) is used with two approaches:
* [In Fedora](https://fedoraproject.org/wiki/Features/SecureBoot)
  * with Secure Boot, shimx64.efi loads grubx64.efi, which verifies it against Fedora's certificate (embedded in the shim). It then loads the kernel in secure mode. During the runtime, the kernel (aware of the fact that it was loaded with SecureBoot) will not load any unsigned modules and userland's DMA.
  * without Secure Boot, shimx64.efi loads grubx64.efi and no validation is performed from there on.
* [In Ubuntu](https://lists.ubuntu.com/archives/ubuntu-devel/2012-June/035445.html)
  * with Secure Boot, similar to Fedora, except no runtime security check is performed (after ExitBootServices is called)
  * without Secure Boot, same as Fedora.

If those security features are applicable to Pugnix as well, it would be nice if we adapt both Fedora's and Ubuntu's approaches and provide user an (`ops-build`-time) options to select characteristics they want for the resultant image: e.g. `--secure-boot` and `--hardened-mode`.

**Comment by francescolavra** [comment]
Yes, this issue should be re-purposed for UEFI support

**Comment by am11** [comment]
another data point: [AMD SEV](https://github.com/AMDESE/sev-tool) and [Intel SGX](https://github.com/intel/linux-sgx) require UEFI for confidential computing in cloud.
of course they also require their respective interfaces to be implemented by the OS, but UEFI is a prerequisite.

**Comment by francescolavra** [comment]
Closed by #1443.
